### ISSUE TITLE
Syncthing-GTK causes high CPU loads

### ISSUE THREAD

**Issue by Schroedingers-Cat** [syncthing/syncthing-gtk#66]
I use syncthing-gtk for several computers with Windows 7 and Ubuntu Gnome. I found out that the process syncthing consumes considerably more CPU when using syncthing-gtk. Without syncthing-gtk, the CPU utilization of syncthing is far more better. The behaviour is documented here: https://discourse.syncthing.net/t/high-cpu-load-idle-when-using-syncthing-gtk/1580

**Comment by kozec** [comment]
Well, Syncthing-GTK is pooling for events every second, so this is not exactly unexpected. I can probably make pooling interval configurable for slower PCs, but UI may feel sluggish with that.

If you can, please, check following:
- Check if you get increased CPU usage without ST-GTK, while WebUI is open. That may indicate problem with Syncthing itself or just not powerfull enough machine.
- syncthing CPU usage may be high right after ST-GTK start, but should decrease shortly and remain near zero after few seconds. If it doesn't, there may be something keeping it repeating same request(s). Try to run `syncthing-gtk -d` to display debug log and check Syncthing daemon output for anything suspicious.
- CPU usage should go down when ST-GTK window is hidden (refresh interval is automatically increased)

I can't reproduce problem on my current machine, but I have Ubuntu Gnome installation at home, so I'll check what could go wrong later this evening.

**Comment by kozec** [comment]
(woo, i can add custom labels :) )

**Comment by AudriusButkevicius** [comment]
What do you mean by `poll for events every second`?

**Comment by kozec** [comment]
@AudriusButkevicius Every time when Syncthing-GTK recieves response to `/rest/events` call, it waits 1s and does same request again. Waiting time is incereased to 5s when ST-GTK is minimized.

This should be same thing what WebUI (or rather EventController) does.

**Comment by AudriusButkevicius** [comment]
Yeah, what about other values that are not part of the event interface?

**Comment by kozec** [comment]
@AudriusButkevicius Well...
- `/rest/errors`, `/rest/connections` and `/rest/system` are requested every 5/25s
- `/rest/completion` is requested for all connected devices after `/rest/connections` response and for particular device, when 'DeviceConnected' is handled. 
- And `/rest/model` is called more or less randomly, but frequently only while Syncthing is transfering some files.

`/rest/completion` is probably most CPU heavy, as it can be requested multiple times at once. But even that shouldn't cause 100+% load unless daemon does something really crazy to compute completion value on each call.

**Comment by AudriusButkevicius** [comment]
Yeah all of the following are expensive:
`/rest/model` - as it has to traverse the database 3 times to figure out the model size for a given folder.
'/rest/completion' - has to traverse the database 2 times to figure out the completion percentage.

(Both of these could potentially be optimized on our side to be honest, traversing the whole keyspace once rather than 2, 3 times)

Well yeah, it is a bit crazy, as we pretty much have to traverse the whole of the database.

**Comment by kozec** [comment]
Oww... So less `/rest/completion` calls, you say :D I can probably disable requesting it if everything is up-to-date. I guess there is no other way how to get same data or, even better, total completion for given device?

**Comment by Zillode** [comment]
Might be related: https://github.com/syncthing/syncthing-inotify/issues/19

**Comment by acidicX** [comment]
I also noticed that syncthing-gtk is consuming an awful lot of CPU.
Currently, I have around 60-70% CPU load on a Core 2 Duo (mobile) 2GHz - constantly. Yeah it is not the newest machine, but plenty fast enough with an SSD for such simple tasks as file syncing ;-)

I guess this could be excused when syncing, but it is even consuming so much CPU time when all other devices are offline. I've set the scan time to 600s to make sure it is no scanning issue. inotify is enabled on all folders though, and one folder is quite large (~23G with 60k files).

Daemon log gives no errors or notices apart from "initial sync complete" for all folders. Normal daemon mode without syncthing-gtk works OK, CPU load drops to zero after a few secs.

Is there a way to get verbose logging from -gtk ?

Version: v0.10.25 / 0.6.2 on Arch x64

**Comment by kozec** [comment]
@Zillode I'll check it just to be 100% sure, but I don't thing that ST-GTK (or rather pyinotify) will descend to symlinked directory.

@acidicX Yes, run it with -d argument. High CPU load comes from ST-GTK directly or fromsyncthing process?

**Comment by acidicX** [comment]
The process with high CPU loads seems to be syncthing, but the ST-GTK debug log shows that it watches a lot of files that are normally ignored (for a reason - they change very often). I guess the CPU load comes from the watcher watching files that it normally should ignore and then somehow triggers the daemon to sync, even if there is nothing to do (because of the ignore pattern).

**Comment by kozec** [comment]
Yeah, watcher doesn't do any checks for ignore patterns. It just reports to syncthing every time when some file in watched directory is changed (with, I think, 5 second cooldown.) That shouldn't be problem normally, but if you have lot of ignored files being changed, it will cause lot of bogus `/rest/scan` calls.

Does high CPU load dissapear if you disable filesystem watching?

**Comment by acidicX** [comment]
Yep, seems to be good. Is the watcher a part of inotify? Should I open a bug there to make sure the watcher respects ignored files?
Also, why does it trigger a scan and does not pass along the file? Not supported in the API yet?

By the way, after enabling/disabling inotify for a directory, the GTK UI only wants to restart the daemon. To truely make it happen, you need to restart the whole ST-GTK though? At least the watcher was still active after I disabled inotify for the folder and just did the daemon restart. Only after a full ST-GTK restart the watcher was gone.

**Comment by kozec** [comment]
Nope, watcher.py is just name of module in ST-GTK. It reports exact path to file that was changed, but I always kinda assumed that daemon will do nothing if that file is ignored. I'll check just in case.

Watching not being removed is bug, I'll look into it, along with not sending ignored files. There is API to get them, so it shouldn't be hard.

**Comment by acidicX** [comment]
Great, thanks for your work!

**Comment by jaystrictor** [comment]
I have got the same problem: Syncthing is consuming an awful lot of CPU when Syncthing-GTK is running.
I disabled watcher.py and I have hidden the Syncthing-GTK window, but the issue still remains.
Every 5 secondy I get a peak of high CPU usage.

**Comment by kozec** [comment]
I already have some code ready, where `/rest/completion` is pulled only if remote index update is detected. It helps, but updates remote device state only after file is downloaded. Now I'm trying to find way how to detect that remote node is being synced before that.

**Comment by AudriusButkevicius** [comment]
Just to let you know that we are also putting effort into making these calls less expensive.

**Comment by jaystrictor** [comment]
@kozec Your patch helps a little bit in the sense that now peaks are only occurring every 10 seconds. But still the CPU fan is going up after like 30 seconds of running Syncthing-GTK.
Question: Why is Syncthing-GTK querying Syncthing *at all* when hidden?

**Comment by kozec** [comment]
@jaystrictor because of notification icon. When hidden, it queries `/rest/events` every 5s and `/rest/errors`, `/rest/connections` and `/rest/system` every 25s. That is, when syncthing daemon is idle. But nothing from that should cause high CPU usage...

**Comment by jaystrictor** [comment]
Ahh I see.

Well, it **does** cause hight CPU usage:

![syncthing-cpu](https://cloud.githubusercontent.com/assets/4204810/6686053/61467720-cc9f-11e4-999f-ced6c3eb8d91.png)

**Comment by kozec** [comment]
And how does it look when ST-GTK is not running and WebUI is loaded in browser?

**Comment by jaystrictor** [comment]
Then there is exactly the same problem. High CPU load every 5 seconds.

**Comment by kozec** [comment]
That's what I was afraid of. I can make refresh rates configurable, so one can choose between having UI updated ASAP and not having CPU fried alive, but ST-GTK can't live without any pulling at all.

@AudriusButkevicius would it be possible to add events for errors? Also, how resource-heavy is `/rest/connections` call?

**Comment by AudriusButkevicius** [comment]
event for errors, you mean puller errors?

I don't think any of the calls you listed is very expensive.

**Comment by kozec** [comment]
> event for errors, you mean puller errors?

Currently ST-GTK needs to pull `/rest/errors` every now and then. Having error event would be more convenient.

**Comment by AudriusButkevicius** [comment]
no, I am sure your cpu hog is completion etc.

**Comment by kozec** [comment]
> no, I am sure your cpu hog is completion etc.

But with commit up there, `/rest/completion` is called only when new files are added (local or remote), so it shouldn't be cause for regular 10s peaks.

Anyway, I'll try to do some profiling after I'll get home this evening.

**Comment by AudriusButkevicius** [comment]
I guess you should just verbosely print all the API calls you make and how long they take to see if there is a pattern.

**Comment by kozec** [comment]
^^ another that should help at least a little.

@AudriusButkevicius I'll try that, thanks.

**Comment by jaystrictor** [comment]
I just noticed that when all repositories are "Up to Date", neither using Syncthing-GTK nor WebUI result in high CPU usage.
It is only when Syncthing is actually syncing that the CPU hog is active.

**Comment by Zillode** [comment]
That's correct, the ```/rest/completion``` is more complex when files are needed, therefore the CPU load will increase. We are working towards a fix, but I think we'll need to drop the (very) precise information about the state of the other devices. See https://github.com/syncthing/syncthing/issues/867

**Comment by kozec** [comment]
With new `FolderSummary` and `FolderCompletion` events (syncthing/syncthing#1555), this should be fixed in ST-GTK version 0.7. Or rather as good as ever can be.
